### PR TITLE
Add chapter describing the GCAM solver.

### PR DESCRIPTION
**Pull request by rplzzz**

### CODE REVIEW

File: solver.md [hunk 1]
```diff
@@ -0,0 +1,461 @@
+---
+layout: index
+title: GCAM Solver
+prev: 
+next:
+current-version: v4.2 
+---
+
+At each time step, GCAM searches for a vector of prices that casue all
```
Comment (bpbond) [inline solver.md]: "cause"

File: solver.md [hunk 2]
```diff
@@ -0,0 +1,461 @@
...
+[solver configuration file](#Solver-Configuration-File).
+
+Solver Algorithms
+-----------------
+
+### Bisection Solver
+
+#### Description
+
+The bisection solver is a multidimensional generalization of of
```
Comment (bpbond) [inline solver.md]: ~~of~~

File: solver.md [hunk 3]
```diff
@@ -0,0 +1,461 @@
...
+change sign at different places along the path, none of which will be
+the solution we seek. Thus, in general, bracketing and bisection is
+not a viable strategy for solving a system of equations.
+
+Despite this shortcoming, bisection can often be used to get the
+solver into the general vicinity of the solution, from which one of the
+other solvers can easily find it. For initial guesses far from
+the solution, bisection can often be faster than more sophisticated
+solution techniques. Moreover, bisection places no conditions on the
+Jacobian matrix, making it useful for jostling the solver out of
```
Comment (bpbond) [inline solver.md]: These paragraphs are interesting and well-written. Here, perhaps briefly define (not so much mathematically, as its role) "Jacobian matrix" for the reader.

File: solver.md [hunk 4]
```diff
@@ -0,0 +1,461 @@
...
+desired solution, while rejecting large steps that take us further
+away from it. Computationally we implement this procedure by scaling
+the iteration step by a factor $$ \lambda \leq 1$$ . We start
+each iteration with $$ \lambda = 1$$ and evaluate $$ f$$ . If it has
+failed to decrease sufficiently, we decrease $$ \lambda$$ until we
+obtain an acceptable step. The heuristic for accepting a step and the
+procedure for decreasing $$ \lambda$$ are described in *Numerical
+Recipes*.<sup>[1](#NR)</sup>
+
+Because the Broyden solver is the only solver algorithm that can drive
```
Comment (bpbond) [inline solver.md]: **Bold** this sentence?

File: solver.md [hunk 5]
```diff
@@ -0,0 +1,461 @@
...
+
+The backtracking procedure greatly enhances the convergence properties
+of the solver over the basic procedure, but the solver will still fail
+under some circumstances. The most commonly occurring such
+circumstance is a singular Jacobian caused by one or more prices
+having no impact on the excess demand in any of the solved markets. In
+GCAM this situation occurs predictably at high and low price
+extremes. Many of the input supply and demand functions saturate in
+extreme price regimes, leading to zero derivatives. One of the ways
+GCAM deals with this problem by excluding markets with obvious price
```
Comment (bpbond) [inline solver.md]: "is by"

File: solver.md [hunk 6]
```diff
@@ -0,0 +1,461 @@
...
+<a name="General-Solver-Parameters"></a>
+
+### General Solver Parameters
+
+* **calibration-tolerance**: During the calibration periods, the
+  tolerance for determining whether or not the calibration was
+  successful.
+* **max-model-calcs**: The maximum number of times the full model will
+  be evaluated in the course of the solution. If the solver has not
+  found a valid solution by that time, it exits and issues a warning.
```
Comment (bpbond) [inline solver.md:288]: Just curious–why only a warning? This isn't fatal?

File: solver.md [hunk 7]
```diff
@@ -0,0 +1,461 @@
...
+
+* **all**: True for any market.  
+* **market-name**: True if the name of the market matches the name
+  given in the filter.  
+* **market-type**: True if the type of the market matches the type
+  given in the filter. The available market types are Normal,
+  Calibration, Inverse-Calibration, Tax, RES, Subsidy,
+  [Trial-Value, Demand, and Price](Cycle-breaking-in-GCAM.html).  
+* **solvable**: True for markets that should be solved. "Solvable" in
+  this context refers to intent, not capability. A market is
```
Comment (bpbond) [inline solver.md:314]: > "Solvable" in this context refers to intent, not capability.

Pretty confusing terminology.

File: solver.md [hunk 8]
```diff
@@ -0,0 +1,461 @@
...
+requirements). The second would exclude the global crude oil, natural
+gas, and coal markets and accept all other solvable markets that are
+not yet solved.
+
+## Advanced Solver Details
+
+
+### Compile-time configuration options
+
+In order to compute its iteration step, the Broyden solver must solver
```
Comment (bpbond) [inline solver.md]: "solve"

File: solver.md [hunk 9]
```diff
@@ -0,0 +1,461 @@
...
+iteration steps that are badly corrupted by roundoff error.  Sometimes
+the algorithm can recover from such a situation, and sometimes not,
+but either way it wastes a lot of time.
+
+To combat this effect, we developed a version of the solver that uses
+Singular Value Decomposition (SVD).  The idea behind SVD is to use
+more sophisticated linear algebra techniques to detect and filter out
+components of $$\delta \vec x$$ that are corrupted by ill-conditioning
+in the Jacobian matrix.  SVD is a more complex algorithm than L-U and
+takes longer to compute, but the iterations save can more than offset
```
Comment (bpbond) [inline solver.md]: "saved"

**Comment by bpbond** [comment]
@rplzzz Overall this is well written and extremely interesting. See above for a few minor suggestions for typos and clarity.

File: solver.md (hunk 7 above)
Comment (rplzzz) [inline solver.md:314]: Agreed, but that's the terminology that GCAM has been using since before I got here.

File: solver.md (hunk 3 above)
Comment (rplzzz) [inline solver.md]: Replaced with "partial derivatives", since that gets closer to the heart of the matter.

File: solver.md (hunk 6 above)
Comment (rplzzz) [inline solver.md:288]: Not currently, although at least one user has asked for a configuration option to make it so.  The rationale behind why solver failures aren't fatal is that sometimes you get failures in which the only markets out of balance are ones that make tiny contributions to the global picture, and continuing gives users the option to consider the solution "good enough" if the major markets are solved.  Reasonable people might differ over whether that's a sensible way to handle the problem.

Certainly, part of what's lacking is that the solution-floor is kind of a blunt instrument.  We would like to exclude from the solution criteria markets that are so small that it isn't reasonable to insist on meeting the tolerance criteria in relative terms for those markets.  However, what constitutes "small" depends on the units of the output.  Notably, some markets are really "markets" that encapsulate consistency conditions, and their natural scale can be _very_ different from the normal markets that track physical goods.

**Comment by rplzzz** [comment]
Thanks for the feedback, Ben.

**Comment by rplzzz** [comment]
Last call for comments on this chapter.  If nobody has anything else to add, I'm going to merge it this afternoon.

**Pull request merged**